plugins/flytekit-ray/flytekitplugins/ray/models.py

```diff
@@ -1,10 +1,170 @@
+    UPLOAD_MODE_EAGER = _core_task.IOStrategy.UPLOAD_EAGER
+    UPLOAD_MODE_ON_EXIT = _core_task.IOStrategy.UPLOAD_ON_EXIT
+    UPLOAD_MODE_NO_UPLOAD = _core_task.IOStrategy.DO_NOT_UPLOAD
+
+    def __init__(
+        self,
+        download_mode: _core_task.IOStrategy.DownloadMode = DOWNLOAD_MODE_EAGER,
+        upload_mode: _core_task.IOStrategy.UploadMode = UPLOAD_MODE_ON_EXIT,
+    ):
+        self._download_mode = download_mode
+        self._upload_mode = upload_mode
+
+    def to_flyte_idl(self) -> _core_task.IOStrategy:
+        return _core_task.IOStrategy(download_mode=self._download_mode, upload_mode=self._upload_mode)
+
+    @classmethod
+    def from_flyte_idl(cls, pb2_object: _core_task.IOStrategy):
+        if pb2_object is None:
+            return None
+        return cls(
+            download_mode=pb2_object.download_mode,
+            upload_mode=pb2_object.upload_mode,
+        )
+
+
+class DataLoadingConfig(_common.FlyteIdlEntity):
+    LITERALMAP_FORMAT_PROTO = _core_task.DataLoadingConfig.PROTO
+    LITERALMAP_FORMAT_JSON = _core_task.DataLoadingConfig.JSON
+    LITERALMAP_FORMAT_YAML = _core_task.DataLoadingConfig.YAML
+    _LITERALMAP_FORMATS = frozenset([LITERALMAP_FORMAT_JSON, LITERALMAP_FORMAT_PROTO, LITERALMAP_FORMAT_YAML])
+
+    def __init__(
+        self,
+        input_path: str,
+        output_path: str,
+        enabled: bool = True,
+        format: _core_task.DataLoadingConfig.LiteralMapFormat = LITERALMAP_FORMAT_PROTO,
+        io_strategy: IOStrategy = None,
+    ):
+        if format not in self._LITERALMAP_FORMATS:
+            raise ValueError(
+                "Metadata format {} not supported. Should be one of {}".format(format, self._LITERALMAP_FORMATS)
+            )
+        self._input_path = input_path
+        self._output_path = output_path
+        self._enabled = enabled
+        self._format = format
+        self._io_strategy = io_strategy
+
+    def to_flyte_idl(self) -> _core_task.DataLoadingConfig:
+        return _core_task.DataLoadingConfig(
+            input_path=self._input_path,
+            output_path=self._output_path,
+            format=self._format,
+            enabled=self._enabled,
+            io_strategy=self._io_strategy.to_flyte_idl() if self._io_strategy is not None else None,
+        )
+
+    @classmethod
+    def from_flyte_idl(cls, pb2: _core_task.DataLoadingConfig) -> "DataLoadingConfig":
+        if pb2 is None:
+            return None
+        return cls(
+            input_path=pb2.input_path,
+            output_path=pb2.output_path,
+            enabled=pb2.enabled,
+            format=pb2.format,
+            io_strategy=IOStrategy.from_flyte_idl(pb2.io_strategy) if pb2.HasField("io_strategy") else None,
+        )
+
+
+class K8sPod(_common.FlyteIdlEntity):
+    def __init__(
+        self,
+        metadata: K8sObjectMetadata = None,
+        pod_spec: typing.Dict[str, typing.Any] = None,
+        data_config: typing.Optional[DataLoadingConfig] = None,
```

Comment:
drop this. Also why add this K8s pod. We dont need that, as Ray config will simply float like a json. So just use K8s pod object?

Reply:
we can keep the k8s pod too, i see that you are actually setting the pod properties. 
This way we could also use pod template?

Reply:
Yeah we had a discussion about this on the backend change here: https://github.com/flyteorg/flyte/pull/5933#discussion_r1820096649

It started as just adding support for resources but we realized it would be more flexible if we added support for something similar to a pod template since that is ultimately what the [kuberay contract](https://ray-project.github.io/kuberay/reference/api/#headgroupspec) is.